Add new collector for InfiniBand statistics

roclark opened the pull request:
Add new metrics for the InfiniBand network protocol including the amount of packets sent and received, the number of times the link has been downed and how many times the link has recovered from an error state.

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

roclark commented:
@joehandzik let me know what you think! I just verified on one node with InfiniBand and another without IB that it works as intended.

joehandzik commented on collector/infiniband.go:
```diff
@@ -0,0 +1,152 @@
+	"github.com/prometheus/common/log"
+)
+
+const infinibandPath string = "class/infiniband"
```
They'll want to see if you can move this into a constructor instead of leaving it as a global variable. Take a look at my set of refactoring here to see what you should look to do: https://github.com/prometheus/node_exporter/pull/432

joehandzik commented on collector/infiniband.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, err
+	}
+
+	// Extract just the filenames which equates to the device names
```
This reads kinda strange, probably just drop the s at the end of equates

joehandzik commented:
You'll want to add a unit test or two to your codebase as well. Again, take a look at the ZFS code or any similar network data code for more guidance.

roclark commented:
@joehandzik, added a couple small unit tests and fixed the other changes you mentioned. Let me know if there is anything else you can see. Thanks!

joehandzik commented on collector/infiniband.go:
```diff
@@ -0,0 +1,154 @@
+	"github.com/prometheus/common/log"
+)
+
+var noIBDevicesFound = errors.New("InfiniBand / No devices detected")
```
I think they're still going to flag these, but if you want to keep them global, you will need to add some sort of prefix that uniquely identifies them. Probably change the naming structure to infinibandNoDevicesFound and infinibandNoPortsFound, based on what they told me with my last PR.

joehandzik commented on collector/infiniband_linux_test.go:
```diff
@@ -0,0 +1,42 @@
+// Copyright 2017 The Prometheus Authors
```
Might be worth considering a test like what exists in the ZFS code: https://github.com/prometheus/node_exporter/pull/432/files#diff-1a2aec762c5ab3c1342874a30dd41671R216 or even something like what the bonding test does: https://github.com/prometheus/node_exporter/blob/master/collector/bonding_linux_test.go

Basically, checking that you're reading the fixtures correctly.

roclark commented:
@joehandzik I expanded the unit tests and moved the two global errors to the collector. Anything else you can see here?

joehandzik commented:
@roclark This all looks really, really good. Please feel free to submit a PR to the node_exporter master branch when you're ready (if you haven't, make sure it's all 'go fmt'-ed).

Commit:
Add new collector for InfiniBand statistics

Add new metrics for the InfiniBand network protocol including the amount of packets sent and received, the number of times the link has been downed and how many times the link has recovered from an error state.

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

Commit:
Add unit tests for InfiniBand collector

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

Commit:
Enable InfiniBand by default

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

Commit:
Update README to include InfiniBand collector

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

Commit:
Update authors after InfiniBand commit

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>
